import{u as m,m as u,R as e,ab as f,bh as p,F as h,w as g,C as b,S as y,r as l,b3 as C,o as q,bi as S,P as v}from"./rpcWorker-CL9qhe4-.js";import{F as w}from"./FeatureDetails-DMfanhiz.js";function E(a,t){var i;return((i=t.tags)===null||i===void 0?void 0:i[a])||t[a]}async function R(a,t){const i=m.getSession(t),{view:n}=t;try{if(n)await n.navToLocString(a);else throw new Error("No view associated with this view anymore")}catch(r){console.error(r),i.notify(`${r}`)}}const d={AM:"The smallest template-independent mapping quality in the template",AS:"Alignment score generated by aligner",BC:"Barcode sequence identifying the sample",BQ:"Offset to base alignment quality (BAQ)",BZ:"Phred quality of the unique molecular barcode bases in the {OX} tag",CB:"Cell identifier",CC:"Reference name of the next hit",CM:"Edit distance between the color sequence and the color reference (see also {NM})",CO:"Free-text comments",CP:"Leftmost coordinate of the next hit",CQ:"Color read base qualities",CR:"Cellular barcode sequence bases (uncorrected)",CS:"Color read sequence",CT:"Complete read annotation tag, used for consensus annotation dummy features",CY:"Phred quality of the cellular barcode sequence in the {CR} tag",E2:"The 2nd most likely base calls",FI:"The index of segment in the template",FS:"Segment suffix",FZ:"Flow signal intensities",GC:"Reserved for backwards compatibility reasons",GQ:"Reserved for backwards compatibility reasons",GS:"Reserved for backwards compatibility reasons",H0:"Number of perfect hits",H1:"Number of 1-difference hits (see also {NM})",H2:"Number of 2-difference hits",HI:"Query hit index",IH:"Query hit total count",LB:"Library",MC:"CIGAR string for mate/next segment",MD:"String encoding mismatched and deleted reference bases",MF:"Reserved for backwards compatibility reasons",MI:"Molecular identifier; a string that uniquely identifies the molecule from which the record was derived",ML:"Base modification probabilities",MM:"Base modifications / methylation ",MQ:"Mapping quality of the mate/next segment",NH:"Number of reported alignments that contain the query in the current record",NM:"Edit distance to the reference",OA:"Original alignment",OC:"Original CIGAR (deprecated; use {OA} instead)",OP:"Original mapping position (deprecated; use {OA} instead)",OQ:"Original base quality",OX:"Original unique molecular barcode bases",PG:"Program",PQ:"Phred likelihood of the template",PT:"Read annotations for parts of the padded read sequence",PU:"Platform unit",Q2:"Phred quality of the mate/next segment sequence in the {R2} tag",QT:"Phred quality of the sample barcode sequence in the {BC} tag",QX:"Quality score of the unique molecular identifier in the {RX} tag",R2:"Sequence of the mate/next segment in the template",RG:"Read group",RT:"Reserved for backwards compatibility reasons",RX:"Sequence bases of the (possibly corrected) unique molecular identifier",S2:"Reserved for backwards compatibility reasons",SA:"Other canonical alignments in a chimeric alignment",SM:"Template-independent mapping quality",SQ:"Reserved for backwards compatibility reasons",TC:"The number of segments in the template",TS:"Transcript strand",U2:"Phred probability of the 2nd call being wrong conditional on the best being wrong",UQ:"Phred likelihood of the segment, conditional on the mapping being correct"},k=u()({compact:{paddingRight:0,paddingTop:0,paddingBottom:0}}),P=["read paired","read mapped in proper pair","read unmapped","mate unmapped","read reverse strand","mate reverse strand","first in pair","second in pair","not primary alignment","read fails platform/vendor quality checks","read is PCR or optical duplicate","supplementary alignment"];function A(a){const{classes:t}=k(),{feature:i}=a,{flags:n}=i;return e.createElement(f,{...a,title:"Flags"},e.createElement(p,{name:"Flag",value:n}),e.createElement(h,null,P.map((r,o)=>{const s=n&1<<o,c=`${r}_${s}`;return e.createElement(g,{key:c,control:e.createElement(b,{className:t.compact,checked:!!s,name:r,readOnly:!0}),label:r})})))}function F({locString:a,model:t}){return e.createElement(y,{onClick:i=>{i.preventDefault(),R(a,t)},href:"#"},a)}function O({value:a}){const[t,i]=l.useState(!1),[n,r]=l.useState(!1),o=String(a);return o.length>100?e.createElement(e.Fragment,null,e.createElement("button",{type:"button",onClick:()=>{C(o),r(!0),setTimeout(()=>{r(!1)},700)}},n?"Copied to clipboard":"Copy"),e.createElement("button",{type:"button",onClick:()=>{i(s=>!s)}},t?"Show less":"Show more"),e.createElement("div",null,t?o:`${o.slice(0,100)}...`)):e.createElement("div",null,o)}const T=l.lazy(()=>import("./SupplementaryAlignments-YwRzpgYj.js")),M=l.lazy(()=>import("./LinkedPairedAlignments-CV-1bzEc.js")),x=["clipPos","flags"],Q=q(function(a){const{model:t}=a,{featureData:i}=t,n=S(i),r=E("SA",n),{flags:o}=n;return e.createElement(v,{"data-testid":"alignment-side-drawer"},e.createElement(w,{...a,omit:x,descriptions:{...d,tags:d},feature:n,formatter:(s,c)=>c==="next_segment_position"?e.createElement(F,{model:t,locString:s}):e.createElement(O,{value:s})}),r!==void 0?e.createElement(T,{model:t,tag:r,feature:n}):null,o&1?e.createElement(M,{model:t,feature:n}):null,o!==void 0?e.createElement(A,{feature:n,...a}):null)});var L=Object.freeze({__proto__:null,default:Q});export{L as A,R as n};
//# sourceMappingURL=AlignmentsFeatureDetail-BeP0LdO4.js.map
